### ISSUE TITLE
[java] UnusedAssignment false positive with call chains

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#2796]
**Affects PMD Version:** 6.26.0

**Description:**

Call chains where side-effects occur in arguments are not handled by UnusedAssignment. This is a known issue, described here

https://github.com/pmd/pmd/blob/59f1fe251e36ee586d1e91796aab054a98a15872/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnusedAssignmentRule.java#L895-L896

It's hard to make this right on master, because of the grammar for expressions. But this is easily fixed in #2794 

**Code Sample demonstrating the issue:**

```java
class Test {
    int a() {
        int a = 10;
        try {
            // both method calls may throw, so a = 2 and a = 4 reach after the try,
            // but not a = 10
            "".substring(a = 2)
              .substring(a = 4);
        } catch (RuntimeException e) {
        }
        return a;
    }
}
```
We get

```
line 3:	The initializer for variable 'a' is never used (overwritten on line 7)
line 7:	The value assigned to variable 'a' is never used (overwritten on line 8)
```
but only the first warning is relevant, the other one is a FP

**Comment by adangel** [comment]
fixed by #2794
